### PR TITLE
feat: Get and check confirmations required for transfers

### PR DESCRIPTION
**Pull request by Kormick**
Epic: #7643 

## What is the purpose of the change
- Adds request and check of the confirmations required for all transfers between chains

## Testing and Verifying
- Unit tests with mocks

## Documentation and Release Note

  - [x] Does this pull request introduce a new feature or user-facing behavior changes?
  - [] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented?
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [x] N/A

### CODE REVIEW

File: x/bridge/observer/bitcoin/bitcoin.go [hunk 1]
```diff
@@ -84,7 +85,7 @@ func (b *ChainClient) SignalInboundTransfer(ctx context.Context, in observer.Tra
 }
 
 // Returns current height of the Bitcoin chain
```
Comment (keruch) [inline x/bridge/observer/bitcoin/bitcoin.go:87]: Don't forget about godoc!

File: x/bridge/observer/bitcoin/bitcoin.go [hunk 2]
```diff
@@ -84,7 +85,7 @@ func (b *ChainClient) SignalInboundTransfer(ctx context.Context, in observer.Tra
...
-func (b *ChainClient) Height() (uint64, error) {
+func (b *ChainClient) Height(context.Context) (uint64, error) {
```
Comment (keruch) [inline x/bridge/observer/bitcoin/bitcoin.go:88]: Like for the argument without a name!

File: x/bridge/observer/bitcoin/bitcoin.go [hunk 3]
```diff
@@ -93,7 +94,7 @@ func (b *ChainClient) Height() (uint64, error) {
 }
 
 // Returns number of required tx confirmations
-func (b *ChainClient) ConfirmationsRequired() (uint64, error) {
+func (b *ChainClient) ConfirmationsRequired(context.Context, bridgetypes.AssetID) (uint64, error) {
 	return 0, nil
```
Comment (keruch) [inline x/bridge/observer/bitcoin/bitcoin.go:98]: Still returns zero

File: x/bridge/observer/osmosis/client.go [hunk 4]
```diff
@@ -145,6 +149,28 @@ func (c *Client) Account(ctx context.Context, addr sdk.AccAddress) (authtypes.Ba
 	return ba, nil
 }
 
+func (c *Client) ConfirmationsRequired(
+	ctx context.Context,
+	assetId bridgetypes.AssetID,
+) (uint64, error) {
+	req := bridgetypes.QueryParamsRequest{}
+	params, err := c.bridgeClient.Params(ctx, &req)
```
Comment (keruch) [inline x/bridge/observer/osmosis/client.go:157]: nit:
```suggestion
	params, err := c.bridgeClient.Params(ctx, new(bridgetypes.QueryParamsRequest))
```

File: x/bridge/observer/osmosis/client.go [hunk 5]
```diff
@@ -145,6 +149,28 @@ func (c *Client) Account(ctx context.Context, addr sdk.AccAddress) (authtypes.Ba
...
+	if err != nil {
+		return 0, errorsmod.Wrapf(ErrQuery, "bridge/params: %s", err.Error())
+	}
+	idx := slices.IndexFunc(params.GetParams().Assets, func(a bridgetypes.Asset) bool {
+		return a.Id == assetId
+	})
+	if idx == -1 {
```
Comment (keruch) [inline x/bridge/observer/osmosis/client.go:164]: It's better to wrap all magic numbers into constants. There is even a linter for that, which is unfortunately disabled in osmosis
```suggestion
	const idxNotFound = -1
	if idx == idxNotFound {
```

**Comment by czarcas7ic** [comment]
I would appreciate a godoc on more of these methods, I wont block on it though. Im still reviewing but will merge soon

File: x/bridge/observer/observer.go [hunk 6]
```diff
@@ -116,31 +139,69 @@ func (o *Observer) sendOutbound(ctx context.Context) {
...
+			confirmed, err := o.isTxConfirmed(ctx, height, &out)
+			if !confirmed || err != nil {
+				if err != nil {
+					o.logger.Error(fmt.Sprintf("Failed to confirm tx %s", err.Error()))
+				}
 				newQueue = append(newQueue, out)
 				continue
 			}
-			err = o.chains[out.DstChain].SignalInboundTransfer(ctx, out)
+			err = o.chains[out.Tx.DstChain].SignalInboundTransfer(ctx, out.Tx)
```
Comment (czarcas7ic) [inline x/bridge/observer/observer.go:158]: I would prefer just checking error first, then checking !confirmed, the nested if is less understandable

File: x/bridge/observer/observer.go [hunk 7]
```diff
@@ -116,31 +139,69 @@ func (o *Observer) sendOutbound(ctx context.Context) {
...
 			}
 		}
 		o.outTxQueue[srcId] = newQueue
 	}
 }
+
+func (o *Observer) getConfirmationsRequired(ctx context.Context, tx Transfer) (uint64, error) {
+	if tx.SrcChain == ChainIdOsmosis {
+		return 0, nil
+	}
```
Comment (czarcas7ic) [inline x/bridge/observer/observer.go:170]: Question on this, we still would need 1 confirmation for Osmosis right?

File: x/bridge/observer/observer.go [hunk 8]
```diff
@@ -116,31 +139,69 @@ func (o *Observer) sendOutbound(ctx context.Context) {
...
+		return false, errorsmod.Wrapf(
+			err,
+			"Failed to get confirmations required for outbound transfer %s",
+			item.Tx.Id,
+		)
+	}
+	if curHeight < cr+item.Tx.Height {
+		item.ConfirmationsRequired = cr
+		return false, nil
+	}
```
Comment (czarcas7ic) [inline x/bridge/observer/observer.go:205]: I don't understand what is happening here. It feels like 

curHeight < item.ConfirmationsRequired+item.Tx.Height

is the same check as

if curHeight < cr+item.Tx.Height

If its not we need explanations with in line comments and/or godocs

File: x/bridge/observer/osmosis/client_test.go [hunk 9]
```diff
@@ -307,6 +311,97 @@ func TestBroadcastTxSuccess(t *testing.T) {
...
+	)
+	require.NoError(t, err)
+	keyring := keyring.NewInMemory(app.GetEncodingConfig().Marshaler)
+	client := osmosis.NewClientWithConnection(ChainId, conn, keyring)
+	defer client.Close()
+
+	_, err = client.ConfirmationsRequired(ctx, bridgetypes.AssetID{
+		SourceChain: "na",
+		Denom:       "na",
+	})
```
Comment (czarcas7ic) [inline x/bridge/observer/osmosis/client_test.go:401]: What if Denom is valid and SourceChain isn't or vise versa

File: x/bridge/observer/osmosis/osmosis_test.go [hunk 10]
```diff
@@ -308,13 +308,80 @@ func TestListenOutboundTransfer(t *testing.T) {
...
+				},
+			},
+		},
+	}
+	ots.ts.BridgeServer.
+		EXPECT().
+		Params(gomock.Any(), expReq).
+		Times(1).
+		Return(expResp, nil)
+	ots.Start(t, ctx)
```
Comment (czarcas7ic) [inline x/bridge/observer/osmosis/osmosis_test.go:342]: I think I saw this setup logic ~3 times, can we extract to helper function

File: x/bridge/observer/observer.go (hunk 7 above)
Comment (czarcas7ic) [reply x/bridge/observer/observer.go:172]: Hm okay, I guess as long as we define confirmations as "number of blocks after the transaction is in a block" then this works. But for bitcoin, I think when a transaction is first in a block that counts as 1 confirmation right? We shouldnt have differing definitions of what a confirmation is

File: x/bridge/observer/osmosis/client_test.go (hunk 9 above)
Comment (czarcas7ic) [reply x/bridge/observer/osmosis/client_test.go]: I am saying we should add this as a test case

File: x/bridge/observer/bitcoin/bitcoin.go [hunk 11]
```diff
@@ -32,10 +34,10 @@ type ChainClient struct {
 	stopChan           chan struct{}
 	outboundChan       chan observer.Transfer
 	observeSleepPeriod time.Duration
```
Comment (keruch) [inline x/bridge/observer/bitcoin/bitcoin.go:36]: nit: 
```suggestion
	fetchInterval time.Duration
```

File: x/bridge/observer/osmosis/client.go [hunk 12]
```diff
@@ -19,19 +20,23 @@ import (
...
 )
 
 type Client struct {
-	chainId   string
-	keyring   keyring.Keyring
-	grpcConn  *grpc.ClientConn
-	txConfig  client.TxConfig
-	txClient  tx.ServiceClient
-	accClient authtypes.QueryClient
+	chainId      string
```
Comment (keruch) [inline x/bridge/observer/osmosis/client.go:33]: chainId is unused

Comment (keruch) [reply x/bridge/observer/osmosis/client.go:33]: oh, it's used, sorry for that

File: x/bridge/observer/osmosis/client_test.go [hunk 13]
```diff
@@ -152,7 +179,6 @@ func TestSignTxSuccess(t *testing.T) {
 		grpc.WithTransportCredentials(insecure.NewCredentials()),
 	)
 	require.NoError(t, err)
-	defer conn.Close()
 
 	keyring := keyring.NewInMemory(app.GetEncodingConfig().Marshaler)
```
Comment (keruch) [inline x/bridge/observer/osmosis/client_test.go:183]: variable name collides with the package name
```suggestion
	kr := keyring.NewInMemory(app.GetEncodingConfig().Marshaler)
```

File: x/bridge/observer/osmosis/client.go [hunk 14]
```diff
@@ -50,18 +55,21 @@ func NewClient(
 	return NewClientWithConnection(chainId, grpcConn, keyring), nil
 }
 
+// NewClientWithConnection returns new instance of `Client`
+// using provided gRPC connection
 func NewClientWithConnection(
 	chainId string,
 	conn *grpc.ClientConn,
 	keyring keyring.Keyring,
 ) Client {
```
Comment (keruch) [inline x/bridge/observer/osmosis/client.go:64]:
```suggestion
) *Client {
```

Comment (keruch) [reply x/bridge/observer/osmosis/client.go:64]: The same applies for all constructors

File: x/bridge/observer/osmosis/client.go [hunk 15]
```diff
@@ -50,18 +55,21 @@ func NewClient(
...
-		chainId:   chainId,
-		keyring:   keyring,
-		grpcConn:  conn,
-		txConfig:  app.GetEncodingConfig().TxConfig,
-		txClient:  tx.NewServiceClient(conn),
-		accClient: authtypes.NewQueryClient(conn),
+		chainId:      chainId,
+		keyring:      keyring,
+		grpcConn:     conn,
+		txConfig:     app.GetEncodingConfig().TxConfig,
```
Comment (keruch) [inline x/bridge/observer/osmosis/client.go:69]: This line will lead to a cyclic import with the `app` package during wiring. Let's just pass this value as an argument

File: x/bridge/observer/observer.go (hunk 7 above)
Comment (keruch) [reply x/bridge/observer/observer.go:172]: In Bitcoin, the number of confirmations: `currentHeight - txHeight + 1`, so yes, the first block counts as 1. In that case we need to modify the conditions in the `isTxConfirmed` method to have
```go
	if curHeight - item.Tx.Height + 1 < item.ConfirmationsRequired {
		return false, nil
	}
```
And return 1 here

Comment (keruch) [reply x/bridge/observer/observer.go]: Fixed!

**Pull request merged**